src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java

```
@@ -168,49 +173,91 @@ void preprocessIngestDocument(IngestDocument ingestDocument) {
+                        setVectorFieldsToDocument(
+                            dataForInference.getIngestDocumentWrapper().getIngestDocument(),
+                            dataForInference.getProcessMap(),
+                            inferenceResults
+                        );
+                    }
+                    handler.accept(ingestDocumentWrappers);
+                } catch (Exception e) {
+                    updateWithExceptions(ingestDocumentWrappers, e);
+                    handler.accept(ingestDocumentWrappers);
                 }
-                List<?> inferenceResults = results.subList(startIndex, startIndex + dataForInference.getInferenceList().size());
-                startIndex += dataForInference.getInferenceList().size();
-                setVectorFieldsToDocument(
-                    dataForInference.getIngestDocumentWrapper().getIngestDocument(),
-                    dataForInference.getProcessMap(),
-                    inferenceResults
-                );
-            }
-            handler.accept(ingestDocumentWrappers);
-        }, exception -> {
-            for (IngestDocumentWrapper ingestDocumentWrapper : ingestDocumentWrappers) {
-                // The IngestDocumentWrapper might already run into exception and not sent for inference. So here we only
-                // set exception to IngestDocumentWrapper which doesn't have exception before.
-                if (ingestDocumentWrapper.getException() == null) {
-                    ingestDocumentWrapper.update(ingestDocumentWrapper.getIngestDocument(), exception);
+            }, exception -> {
+                try {
+                    for (IngestDocumentWrapper ingestDocumentWrapper : ingestDocumentWrappers) {
+                        // The IngestDocumentWrapper might already run into exception and not sent for inference. So here we only
+                        // set exception to IngestDocumentWrapper which doesn't have exception before.
```

I think here we do updateWithExceptions(ingestDocumentWrappers, e) in multiple catch blocks. But we only consider the case that IngestDocumentWrapper might already run into exception and not sent for inference here which does not make sense. It's possible the other catch block can override the exception of IngestDocumentWrapper incorrectly. e.g. The IngestDocumentWrapper can have exception A and not sent for inference but now we can potentially override its exception related to inference.

I think we should either propagate the exception to one catch block and handle it with the consideration that IngestDocumentWrapper can already run into the exception. Or we should modify `updateWithExceptions` function to account for that.

---

this was added to ensure that if any of the lines while handling the exceptions in for loop throws exceptions, it will be handled properly. I understand that this may not be needed, but was brought up as a possible risk in previous pr

---

When would this situation occur? `The IngestDocumentWrapper might have already encountered an exception and was not sent for inference.` It would be helpful if you could point out the code where the exception is set before inference.

I agree that there's an inconsistency in how we handle this—checking if an exception is already set only in this specific place but not elsewhere.

---

see [link](https://github.com/opensearch-project/neural-search/blob/fc5936c1c8b9d44eabd441c1393b846d0c707b01/src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java#L310-L311)

---

maybe we can propagate the exception to top level and set the exceptions there?

---

I don't think we can propagate the exception. One option is to remove this check and override the exception for all documents. When an exception occurs here, retaining the original exception may not make a significant difference.

---

another option is to keep the check for all exception cases, which will only store the first exception encountered for all documents.

---

this way, we'll know which ingest document failed with which exact exception

---

that would work as well.

---

I wonder do we really check if the doc already run into the exception and skip the embedding generation for it? If that's the case it makes sense to keep the actual exception fail the ingest of the doc which can be helpful for debugging.